src/protocol/ipa_prf/boolean_ops/addition_low_com.rs

```diff
@@ -0,0 +1,264 @@
+pub async fn integer_add<C, XS, YS>(
+    ctx: C,
+    record_id: RecordId,
+    x: &AdditiveShare<XS>,
+    y: &AdditiveShare<YS>,
+) -> Result<AdditiveShare<XS>, Error>
+where
+    C: Context,
+    for<'a> &'a AdditiveShare<XS>: IntoIterator<Item = AdditiveShare<XS::Element>>,
+    YS: WeakSharedValue + CustomArray<Element = XS::Element>,
+    XS: WeakSharedValue + CustomArray + Field,
+    XS::Element: Field,
+{
+    let mut carry = AdditiveShare::<XS::Element>::ZERO;
+    addition_circuit(ctx, record_id, x, y, &mut carry).await
+}
+
+///saturated unsigned integer addition
+/// currently not used, but it is tested
+/// adds y to x, Output has same length as x (we dont seem to need support for different length)
+/// # Errors
+/// propagates errors from multiply
+#[cfg(all(test, unit_test))]
+pub async fn integer_sat_add<C, S>(
+    ctx: C,
+    record_id: RecordId,
+    x: &AdditiveShare<S>,
+    y: &AdditiveShare<S>,
+) -> Result<AdditiveShare<S>, Error>
+where
+    C: Context,
+    for<'a> &'a AdditiveShare<S>: IntoIterator<Item = AdditiveShare<S::Element>>,
+    S: CustomArray + Field,
+    S::Element: Field,
+{
+    use crate::ff::Expand;
+    let mut carry = AdditiveShare::<S::Element>::ZERO;
+    let result = addition_circuit(
+        ctx.narrow(&Step::SaturatedAddition),
+        record_id,
+        x,
+        y,
+        &mut carry,
+    )
+    .await?;
+
+    //if carry==1 {all 1 array, i.e. Array[carry]} else {result}:
+    //compute carry*Array[carry]+(1-carry)*result = result+carry(Array[carry]-result)
+    let carry_array = AdditiveShare::<S>::expand(&carry);
+    let sat = result.clone()
+        + carry_array
+            .multiply(
+                &(carry_array.clone() - result),
+                ctx.narrow(&Step::MultiplyWithCarry),
+                record_id,
+            )
+            .await?;
+    Ok(sat)
+}
+
+///addition using bit adder
+/// adds y to x, Output has same length as x (carries and indices of y too large for x are ignored)
+///implementing `https://encrypto.de/papers/KSS09.pdf` from Section 3.1
+///for all i: output[i] = x[i] + (c[i-1] + y[i])
+/// # Errors
+/// propagates errors from multiply
+async fn addition_circuit<C, XS, YS>(
+    ctx: C,
+    record_id: RecordId,
+    x: &AdditiveShare<XS>,
+    y: &AdditiveShare<YS>,
+    carry: &mut AdditiveShare<XS::Element>,
+) -> Result<AdditiveShare<XS>, Error>
+where
+    C: Context,
+    for<'a> &'a AdditiveShare<XS>: IntoIterator<Item = AdditiveShare<XS::Element>>,
+    XS: WeakSharedValue + CustomArray,
+    YS: WeakSharedValue + CustomArray<Element = XS::Element>,
+    XS::Element: Field,
+{
+    let mut result = AdditiveShare::<XS>::ZERO;
+    for (i, v) in x.into_iter().enumerate() {
+        result.set(
+            i,
+            bit_adder(
+                ctx.narrow(&BitOpStep::from(i)),
+                record_id,
+                &v,
+                y.get(i).as_ref(),
+                carry,
+            )
+            .await?,
```

Comment:
What makes this "low communication" exactly? It seems to me that this perfectly replicates existing code: https://github.com/private-attribution/ipa/blob/main/src/protocol/boolean/saturating_sum.rs#L131 without any additional communication reduction.

This still performs the multiplications per-bit in a loop, without bit-packing. So this will necessarily send one byte of communication per bit being added (just like the code currently does).

Reply:
Yes so this is just the straightforward depth N addition circuit with N bits of communication per party. The underlying structure is a Boolean Array (BA) and not a bit decomposed GF. We could at some point implement the parallel prefix adder to reduce the depth to log N at the cost of 3N bits of communication per party. So there is no communication saving to the previous addition it only changes the underlying structure from a bit decomposed GF to a BA.

When multiplying a single bit in a BA it gets converted into a bool which is also 1 byte. However, a BA is more compact in memory and it can also be send more efficiently to other parties than a bit decomposed GF. For example when revealing it or shuffling it.

Reply:
Resolved: This is low communication compared to another protocol that's not implemented.